### ISSUE TITLE
The UI fonts are inconsistent across the app

### ISSUE THREAD

**Issue by dtcyganok** [koreader/koreader#12817]
#### Issue

There are tons of places where different typefaces are used: menus, ToC, keybord, USB connection screeen, et cetera. Do we really need to use another monospace font for all of these things?
Additionally font sizes are inconsistent across the windows (statistics, dictionary, Toc) as well. There is no any logic in font sizes. At least I can't see it...

As a result the UI feels cluttered and disordered. KOReader is definitely the best alternative e-book software, although there are quite a few annoying peculiarities...

**Comment by poire-z** [comment]
We only use 2 fonts for the UI: Noto Sans (regular or bold), and DroidSansMono for the monospace stuff.
(The USB screen may not be koreader itself, but some intermission screen done with [fbink](https://github.com/NiLuJe/FBInk) (which doesn't use KOReader fonts but some embedded in its binary executable) while koreader has exited, until it is restarted.)

For the sizes, the logic is... we use the one that feels right to display the usual amount of text for the context in the amount of space available (also thinking about translation and small width languages like English and larger width languages like German) - so, yes it varies, You wouldn't want the TOC displayed with the size of the font used for the menu titles, and see your chapter titles truncated, and only 10 toc items per page and the need to navigate next pages... so we went smaller.
(The top window title font size may also vary, depending if there may be a subtitle - ie. the file page in File browser vs. ToC. And so does the size of the icons in the top left and right corners.)

If you want a consistent font size, well... read a book :)

**Comment by dtcyganok** [comment]
Using just 2 typefaces is enough to make a mess, especially if they are used inappropriately. What do I mean by that? Well, the monospace font is overly used. Look at the popular modern GUI. Nobody uses monospace font so often. The only place for a monospace font is something technical, e.g. terminal, text editor, logs viewer, etc. So my suggestion to use monospace font thoughtfully.

Perhaps in some cases using different font  sizes is indeed reasonable (for example to display long metadata in file manager), but in other cases isn't so. Let's look at the book statistics and the book information windows. In the first case font sized changed to fit everything in one screen, in the second case we have pagination. Why? My suggestion is to review font sizes across the app.

Speaking about the ToC, I would call the de facto design decision is controversial. To me the standard font size would be more preferable.

**Comment by Frenzie** [comment]
> Well, the monospace font is overly used. 

Where is it used at all?

**Comment by dtcyganok** [comment]
At a glance I can name a few places. Probably there are a lot more. They are:

- ToC
- Keyboard
- USB connection screen (created by koreader)

I'm not sure that the ToC and the keyboard typefaces are monospace, but they certainly differ from the main UI font.

**Comment by poire-z** [comment]
> I'm not sure that the ToC and the keyboard typefaces are monospace, but they certainly differ from the main UI font.
> The ToC contains 2 typefaces (the right column's font differs).

The long chapter title on the left is not monospace. The thingies on the right are indeed monospace - as with most of the panels using our generic Menu widget: as they display (page) numbers, it's cleaner when digits are aligned (the early developers decided that, and I kinda like it): if you have pags `211` and page `288`, with a monospace font, you get the `2` vertically aligned. With a non-monospace font, you may get the numbers not all the same widths and it would look rugged/ugly (I think that with NotoSans at least, even if not monospace, the digits glyphs have the same width anyway, so it may not really matter - but what if some guy use a monkey patch to use another font which doesn't have that :)).
I guess it's the same idea with the keyboard keys: with monospace letters, things will look more even.

> Speaking about the ToC, I would say the de facto design decision is controversial. To me the standard font size would be more preferable.

Not sure about our defaults (but nobody else complained :)), but for the ToC, you can tweak it to how you like it:
![image](https://github.com/user-attachments/assets/24d79f94-6e20-425c-976e-089b7f4e59f7)
You can do that and more also for the bookmarks/highlights list.

> Let's look at the book statistics and the book information windows. In the first case font sized changed to fit everything in one screen, in the second case we have pagination. Why? My suggestion is to review font sizes across the app.

I'm not sure we have "font sized changed to fit everything in one screen" (we may, dunnno), but for these 2 items and others, which are based on a same generic widget, KeyValuePage, you can decide how many items you want per page (via File browser > first tab > settings):
![image](https://github.com/user-attachments/assets/87ee3de2-8cb8-4794-a996-022fdf3d58bd)
The number of items per page (the font adjusting for the chose amount) is more graspable for a (normal :)) user than some rigid focus on font size consistency :)

**Comment by Frenzie** [comment]
I see that @poire-z posted everything about the variable sized numbers already, but I'll add that things before and after the program (like the USB connection display) look different for technical reasons. It's a different program. Making that slower and more brittle with fancy fonts is probably an anti-goal but @NiLuJe can say if I'm phrasing that overly negatively.

**Comment by dtcyganok** [comment]
> if you have pags 211 and page 288, with a monospace font, you get the 2 vertically aligned.

Yeah, that's cool, no doubt. But UX designers usually choose a sane and modern typeface with monospace numerals, such as `IBM Plex Sans`, `Source Sans`, `Roboto`, and many others. Also one can use `tnum` OpenType feature, to make digits monospaced. The point is not to mix typefaces.

> Not sure about our defaults (but nobody else complained :))

That means nothing. People may not be willing to spend their time trying to convince someone of a reasonable decision. And users are usually not inclined to complain, especially about free software.

And I still don't understand why the two aforementioned similar windows have different font sizes...

> The number of items per page (the font adjusting for the chose amount) is more graspable for a (normal :)) user than some rigid focus on font size consistency :)

That's emotional reasoning. And `The number of items per page` setting is pretty peculiar to me. I believe that the user of a well-designed application never thinks about this at all.

**Comment by NiLuJe** [comment]
> Making that slower and more brittle with fancy fonts is probably an anti-goal but @NiLuJe can say if I'm phrasing that overly negatively.

Nope, that's fairly spot-on: for simplicity's sake, FBInk uses stb for font rendering, so there's no fancy OpenType features support to begin with, and no text shaper at all, so complex scripts are out, too ;).

> choose a sane and modern typeface

Depends on your goals. Ours is "get as as much unicode coverage in a single typeface", not "fancy OpenType features". That pretty much narrows it down to Noto. Which is ubiquitous enough that it doesn't really feel horribly alien to anybody, which is sort of a plus (or a downside, if you want your app like a special snowflake, but we're not Apple).

**Comment by NiLuJe** [comment]
> * Keyboard

This is literally a grid, I fail to understand how using anything else than a monospaced font would make sense in that context.

**Comment by MoreFoxBeans** [comment]
> > * Keyboard
> 
> This is literally a grid, I fail to understand how using anything else than a monospaced font would make any kind of sense in that context.

The monospace font on the keyboard bothers me as well, I used to manually change it in virtualkeyboard.lua but gave up when it would change back every release. If we could get a dedicated keyboard font variable in font.lua (so it can be changed with a user patch) that would be lovely

**Comment by poire-z** [comment]
> And I still don't understand why the two aforementioned similar windows have different font sizes...

Must be because KeyValuePage has a `single_page` option that makes it decide for a smaller font size if needed.
And only 4 things in the statistics plugin use it (including Current stats and Current book stats) - because we know the nb of items should not be greater than 10-20 and its best having it all visible and not have to paginate. Other stuff using KeyValuePage usually display an unknown and possibly large number of items.
I had forgotten about it - and I think Book information could benefit from having it set to true (I'm used to increase my "Items per page" value each time @hius07 adds an item to it :)).

**Comment by poire-z** [comment]
> it seems really trivial to add a monospace toggle to the keyboard appearance menu. 

I'd rather not have a UI toggle for such a thing (we already have a lot, it's not because it's easy to add one that we should :))
Either we switch to non-monospace (CJK or Arabic keyboards probably get to use a non-monospace fallback font), or if most of us prefer the monospace font (I personally don't care), we make it easier to change with a user-patch.

**Comment by Frenzie** [comment]
I think the keyboard looks fine with regular Noto Sans, even if I find the monospace surprisingly charming.

**Comment by poire-z** [comment]
> Must be because KeyValuePage has a `single_page` option that makes it decide for a smaller font size if needed.
> And only 4 things in the statistics plugin use it (including Current stats and Current book stats) - because we know the nb of items should not be greater than 10-20 and its best having it all visible and not have to paginate. Other stuff using KeyValuePage usually display an unknown and possibly large number of items.
> I had forgotten about it - and I think Book information could benefit from having it set to true (I'm used to increase my "Items per page" value each time [@hius07](https://github.com/hius07) adds an item to it :)).

```diff
--- a/frontend/apps/filemanager/filemanagerbookinfo.lua
+++ b/frontend/apps/filemanager/filemanagerbookinfo.lua
@@ -259,6 +259,7 @@ Source (print edition):
     local KeyValuePage = require("ui/widget/keyvaluepage")
     self.kvp_widget = KeyValuePage:new{
         title = self.title,
+        single_page = true,
         value_overflow_align = "right",
         kv_pairs = kv_pairs,
         values_lang = values_lang,
```
Any thought about using this ? @hius07 @Frenzie, others?
I currently get `Notebook file` (only, on my Kobo) and `Review` (on the emulator) on the 2nd page. I would even decrease my "Info list items per page" to 15 so I don't see `Rating` (that I don't care about) on page 1. But then, the font size gets bigger and more stuff gets truncated (so, if single_page=true, needing to put 18 items on page 1, it would decrease the font size, truncating less stuff).
Or do you feel it's better to just not change anything and let users adapt "Info list items per page" to how much they want or need to be readable (I think I've always tuned "Info list items per page" for what I want to see in Book information, it was all for years before new stuff gets added).

**Comment by Frenzie** [comment]
What would happen in landscape mode?

**Comment by hius07** [comment]
I'd let a user to choose.
One-page font size might be too small for poor eyes.

**Comment by poire-z** [comment]
> What would happen in landscape mode?

It would be really small :) (So, this already happens for Statistics' Current stats and Current book stats, but it's usually just numbers, so probably more graspable to have all of them on a single page, even if small, to relate numbers to other ones).

Ok, let's forget about this idea and keep the current stuff.
